### DISCUSSION TITLE
How to get `SSH FS` working?

### DISCUSSION THREAD

**Discussion by TomS-** [coder/code-server#4075] [Q&A] [answered]
https://marketplace.visualstudio.com/items?itemName=Kelvin.vscode-sshfs

Whenever I try to edit the config I get the error "React Dev server not started" on `127.0.0.1:3000` 

Is this a plugin that will not work on the browser version?

**Comment by jsjoeio** [comment]
> Whenever I try to edit the config I get the error "React Dev server not started" on 127.0.0.1:3000

Hmm... I'm not sure how that extension works under the hood 🤔 But if you post steps to reproduce this, I'm happy to see if I get the same thing locally using the [open-vsx version of this extension](https://open-vsx.org/extension/Kelvin/vscode-sshfs)

**Reply by TomS-** [reply]
Yeah of course. 
I installed the code-server using the script in the readme file, did the login process. Installed the SSH FS plugin, went to the new tab created by the plugin. Clicked the "Create a SSH FS configuration` icon and I get the following error:
```
[Extension Host] [WARNING] [WebView] Error connecting to React dev server: Error: connect ECONNREFUSED 127.0.0.1:3000Logged at:	at ClientRequest.<anonymous> (/home/coder/.local/share/code-server/extensions/kelvin.vscode-sshfs-1.21.2/dist/425.extension.js:1:1281)	at ClientRequest.emit (events.js:375:28)	at Socket.socketErrorListener (_http_client.js:475:9)Reported from WebView
notificationsAlerts.ts:40 Could not connect to React dev server. Not running?
onDidChangeNotification @ notificationsAlerts.ts:40
main.js:211 Uncaught (in promise) Error: Service Workers are not enabled in browser. Webviews will not work.
    at main.js:211
    at new Promise (<anonymous>)
    at main.js:209
```

I have also tried it through Docker in case I'm missing any dependencies and I'm still having the same error.

**Reply by jsjoeio** [reply]
Good news: I was able to reproduce

https://user-images.githubusercontent.com/3806031/131414806-8e4f42ba-e831-41a5-9197-a8a26d2bc32a.mov

Bad news: I'm not sure we can do much here to help 🤔

I found where the [error happens](https://github.com/SchoofsKelvin/vscode-sshfs/blob/a41c435e494dac086e7ec21f7733d9dd755f742d/src/webview.ts#L38). 

Usually when you run a React app in code-server, you have to make a few adjustments to get it working (see [docs](https://coder.com/docs/code-server/v3.11.1/guide#proxying-to-create-a-react-app)) so my hunch says you'd have to make those changes in the extension to get it to work, but maybe not. Wondering if @code-asher has any thoughts?

**Reply by TomS-** [reply]
Thank you for your fast reply! I'm able to manually change the config file and the extension works as expected. It only seems to be the config screen that is effected.

**Reply by jsjoeio** [reply]
Ahhh got it. Well looking at the original code, it seems like it's using React to render the webview? And because of that, the webview won't load. There could be another webview issue in code-server (we've had that in the past).  But yeah, my hunch is you'd need to modify the settings for the React app in the extension so that it can work with code-server's proxy

**Comment by code-asher** [comment, marked as answer]
Looks like the extension is trying to connect to 127.0.0.1 from the web view which will not work since that is your local machine and not where code-server is running.

The extension will need to be modified to use code-server's proxy I think. The `vscode-R` extension has a similar problem, see https://github.com/cdr/code-server/issues/1510.

**Reply by code-asher** [reply]
Or like @jsjoeio said if the extension has a place to configure the endpoint that would also work!

**Reply by SchoofsKelvin** [reply]
Th reason the extension tries to connect to localhost is because by default, the Settings UI uses a pre-built React project that's included in the extension. But for easier development, when the extension is run in non-production mode, it would first try to connect to the local React dev server (if running), otherwise fallback to the included build.

I've noticed two issues with `code-server` that don't seem to happen in the regular VS Code (or users of the OpenVSX release):
- `ExtensionContext.extensionMode` (passed to `activate`) is set to `ExtensionMode.Development` (integer value of 2) instead of `Production`. In other words, `code-server` runs extensions in non-production mode, which my extension took as a sign to enable that devserver shortcut. Not really a problem, if it weren't for the next problem:
- The fallback fails to load for some people (and once for me right after I installed `code-server` and the extension). Not entirely sure why, but my guess is `code-server` not (always) properly proxying `webview.asWebviewUri(...)`-produced URLs, resulting in the JS scripts not loading. _If I can reproduce the blank screen (which I personally only got the first time I tried opening my Settings UI), I might be able to figure out more._

**Reply by jsjoeio** [reply]
I can't answer the `ExtensionContext.extensionMode` point (maybe @code-asher or @TeffenEllis can though), but for the second point, what version of code-server are you using? We had an issue with the `asWebviewUri` after merging a patch, but that should all be fixed in the latest version of code-server (3.12.0).

**Reply by SchoofsKelvin** [reply]
I tested it using 3.12.0 on Ubuntu which I installed about an hour or two ago, so should be the latest version. As for the other users of choofsKelvin/vscode-sshfs#279, I'm not sure. The first report was on August 4, while on September 2 someone still reported suffering from a blank webview.

For the `ExtensionContext.extensionMode` issue, I've currently changed the extension a bit, having it instead depend on a certain environment variable to enable debug mode. While it "solves" (in the eyes of the users) the unnecessary error, some users will still suffer from the blank webview issue.

**Reply by jsjoeio** [reply]
Ah got it. Well, thanks for that addition context. Reading through https://github.com/SchoofsKelvin/vscode-sshfs/issues/279 was helpful.

I'm unsure about the blank webview then. Are there any logs in the browser when the blank webview happens? 

Actually...I'm going to comment on your issue in SSH-FS instead. That might be easier to track things.

**Reply by code-asher** [reply]
Ahh very enlightening! I bet there is some production flag we are not setting so it defaults to development.

**Reply by jsjoeio** [reply]
Awesome. I opened an issue for us to fix this: https://github.com/cdr/code-server/issues/4219

Thanks @SchoofsKelvin!

**Reply by benz0li** [reply]
This [production flag missing] may be the reason for https://github.com/REditorSupport/vscode-R/issues/275#issuecomment-929253126, too; the 'data viewer'  of the R extension not working correctly.

@jsjoeio @code-asher Maybe it's worth having a look at https://vscode-r.jupyter.b-data.ch, where everything is in place.
ℹ️ Sign in with your GitHub account which is whitelisted.

```
> R: Create R terminal
```

```R
View(mtcars)
```

**Reply by jsjoeio** [reply]
Thanks @benz0li! I took a quick look but I'm not seeing anything in the console that looks suspicious so I'm not sure 🤔


https://user-images.githubusercontent.com/3806031/135155948-9974b093-07b1-4f7b-b819-ba795a6a5a4a.mov

**Reply by benz0li** [reply]
![Screenshot 2021-09-29 at 06 26 23](https://user-images.githubusercontent.com/20233085/135203314-7f52e365-0913-4370-9ba0-0bb2377a64a2.png)

**Reply by benz0li** [reply]
[console-export-2021-9-29_6-29-20.txt](https://github.com/cdr/code-server/files/7248830/console-export-2021-9-29_6-29-20.txt)

**Reply by benz0li** [reply]
Links in [Markdown] previews point to the wrong location, too:

![Screenshot 2021-09-29 at 07 02 01](https://user-images.githubusercontent.com/20233085/135206415-381001ef-9f89-4d7a-8093-f422d9266560.png)

I have this issue since upgrading to v3.12.0; links in previews worked perfectly well with v3.10.2. I didn't use v3.11.x because I couldn't get the Python extension working.

**Reply by jsjoeio** [reply]
Thanks for the screenshots and the console export. Some of the errors look familiar. Do you mind opening an issue? That will be easier for us to track (vs this Discussion) and then we can take a look.

**Reply by benz0li** [reply]
Fixing https://github.com/cdr/code-server/issues/4219 should resolve the WebView and [wrong links in] Preview issue.

Nothing more to add.

**Reply by jsjoeio** [reply]
Awesome! I'll add "high priority" to that and see if we can get it into the next version milestone!
